Upgrade/Downgrade Path for 2i Bigint Bug

jrwest opened the pull request:
As reported in #499, 2i range queries involving bigints could return incorrect results. The issue was caused by https://github.com/uwiger/sext/issues/5. The fix introduces an incompatibility in sext's encoding of bigints. This PR addresses handling of both the fixed and legacy encoding in Riak and provides a path for 2i users to upgrade stored indices that may be using the legacy format. A path for downgrading is also provided. 

With this change, when first started, Riak will write new indicies (or update indicies) using the legacy format. When indicies are removed (when an object is deleted or updated), initially, the backend will attempt to remove both the legacy and fixed encoding of the index. After `riak_kv_util:fix_incorrect_indices/0,1` has been run the additional delete of the legacy index will no longer occur. The additional delete will also not occur in partitions that are empty on startup. 

Users wishing to downgrade can use `riak_kv_util:fix_incorrect_indices/2` passing the option `{downgrade, true}`. This performs a similar operation to upgrading, however, after starting, index writes will be encoded in the legacy format.

In the case `riak_kv_util:fix_incorrect_indices/0,1` crashes or returns errors it is expected that it will be re-run. Re-running will "continue from where it left off" in the sense that indices rewritten before the crash will not be rewritten again. Results are written to the log as `info` messages.

Additionally, this PR changes the dependency on `sext`. It bumps the version to 1.1, which is necessary to include the encoding fix and uses Ulf Wiger's repository instead of ESL's since Ulf's is now the official one.

engelsanchez committed:
support for handling legacy sext encoding of 2i keys

The version of sext used by riak 1.3 and prior did not
preserve the sort order of bigints when serializing erlang
terms resulting in 2i queries that returned missing or
invalid results (riak_kv#499). The updated version of sext fixes the issue
with an incompatible change to the encoding of bigints. Special
handling of this incompatability is necessary to provide a path
for indexes to be fixed (as they are written to disk).

A utility function for reformatting the on-disk indexes (per-node)
is provided. This allows indexes to be upgraded on a running node.
The utility folds over all indexes, per partition, and rewrites any
invalid index (those containing bigints encoded using the old version
of sext). In order to safely migrate indexes, until the fold that
rewrites the indexes in the new format completes, all riak_object
puts/deletes must delete the legacy index format in addition to the new format

Upon successful completion or if the partition is empty a flag is
written to the leveldb backend to ensure that future writes/deletes
to not inccur the additional cost of deleting the legacy index format

engelsanchez committed:
Fix eleveldb fold_objects to start at 1st object

Also fixed special bucket option so it starts at the beginning of the
given bucket. But that doesn't appear to be used anywhere...

jrwest committed:
add support for downgrading back to legacy index format

jrwest committed:
Merge pull request #503 from basho/eas-fix-eleveldb-fold-objects-first-key

Fix eleveldb fold_objects to start at 1st object

jrwest committed:
cleanup a few comments

jrwest committed:
Merge branch 'jrw-eas-2i-fix-keys-backport' of github.com:basho/riak_kv into jrw-eas-2i-fix-keys-backport

jrwest commented:
accompanying `riak_test` PR: https://github.com/basho/riak_test/pull/234

engelsanchez commented on src/riak_kv_eleveldb_backend.erl:
```diff
@@ -51,6 +54,7 @@
 
 -define(API_VERSION, 1).
 -define(CAPABILITIES, [async_fold, indexes]).
+-define(FIXED_INDEXES_KEY, sext:encode(fixed_indexes)).
```
I've been thinking about this and wondering: we are going to live with these markers for all eternity. Should we maybe make something more compatible with existing keys like:
- Objects are {o,Bucket, Key}
- Index data is {i,Bucket,Field,Term,Key}
- Meta-data {md, Key}
  To accommodate similar markers in the future without making a mess that will need maintenance forever. What do you think?

jrwest replied:
i actually chose to not use a tuple for that exact reason but maybe that was my mistake. As my reasoning went, using an atom guarunteers `FIXED_INDEX_KEY` is always less than (in erlang term order, and thus sext) any tuple term.

EDIT: "using an atom"

engelsanchez replied:
This is a very small issue. The order doesn't really matter that much. Actually, a 2 tuple will always sort before a 3 tuple anyway when using sext. I'm just thinking that down the line we might want more markers and people will get creative with the names. So maybe adding the concept of keys that are global meta-data for the partition might make that less likely.  That way we could avoid even more special code by having some to_md_key function and such.

jrwest replied:
if you'd really like that cleanup to be made i have no objection. personally, i would make that change when we actually ran across the issue (its easy to still write a `to_md_key` that works on atoms and 2-tuples).

engelsanchez replied:
I'm thinking of to_md_key(fixed_indexes), to_md_key(some_other_crazy_marker) that map to {md,Key}. Instead of then having sometimes ?FIXED_INDEXES_KEY + another generic path for to_md_key(some_other...). That is, creating a new namespace that can be used for all markers of this kind.  We could also just have atoms I suppose, making this key compatible and other markers in the future being plain atoms, but then someone comes along and adds a marker that's a tuple... I think wrapping it up in {md, <marker key>} would be more future proof.

engelsanchez commented on src/riak_kv_eleveldb_backend.erl:
```diff
@@ -159,13 +191,74 @@ put(Bucket, PrimaryKey, IndexSpecs, Val, #state{ref=Ref,
+index_deletes(FixedIndexes, Bucket, PrimaryKey, Field, Value) ->
+    KeyDelete = [{delete, to_index_key(Bucket, PrimaryKey, Field, Value)}],
+    LegacyDelete = [{delete, to_legacy_index_key(Bucket, PrimaryKey, Field, Value)}
+                    || FixedIndexes =:= false],
```
How about avoiding adding LegacyDelete if the result of to_index_key and to_legacy_index_key is the same? One less possibility of a performance regression right there.

jrwest replied:
ah good call, so that way we dont issue the deletes for string indices and the like. will make that change

engelsanchez commented on src/riak_kv_vnode.erl:
```diff
@@ -511,7 +511,57 @@ handle_command(?KV_VNODE_STATUS_REQ{},
+                {error, Reason}
+        end,
+    {reply, Reply, State};
+handle_command({get_index_entries, ForUpgrade},
```
Not that it matters for 1.3.1, but this names looks like a generic index scan instead of a scan returning entries to upgrade. What's in a name? :)

jrwest replied:
haha i changed it b/c now it fixes and "unfixes" indexes. naming is hard. realize i didn't continue that up the list of commands though. may just change it back while im in there to be consistent

engelsanchez commented on src/riak_kv_vnode.erl:
```diff
@@ -511,7 +511,57 @@ handle_command(?KV_VNODE_STATUS_REQ{},
+               State=#state{mod=Mod,
+                            modstate=ModState0}) ->
+    case Mod of
+        riak_kv_eleveldb_backend ->
```
Oh shite, we have to support multi backend. When I wrote this I didn't intend for it to live for too long.  More work!

jrwest replied:
hmm multibackend will be interesting. IIRC the multibackend only supports indexes when all "sub-backends" support it [1]. 

[1] https://github.com/basho/riak_kv/blob/master/src/riak_kv_multi_backend.erl#L104

engelsanchez committed:
Add key namespace for special leveldb markers

My argument is that this may make it less messy down the road if others
start adding other special markers in leveldb partitions. They now have
a separate namespace, like objects and index keys do.

jrwest committed:
add console support for reformatting indices

engelsanchez committed:
Fix 2i bigint eq query during transition

This causes a double scan while in the transition from the legacy
(broken for big int) sext encoding for 2i keys. Once starting at the old
key value, later at the new one. Once the upgrade or
downgrade is finished, it should return to a normal scan.

jrwest committed:
add support for multi-backend index reformatting

the multibackend supports indexing on buckets assigned to sub-backends
that support indexing. LevelDB sub-backends must be reformatted.

jrwest committed:
don't delete same index key twice during legacy delete

russelldb commented on src/riak_kv_eleveldb_backend.erl:
```diff
@@ -99,11 +105,30 @@ start(Partition, Config) ->
+            {ok, State#state{fixed_indexes=true}};
+        false ->
+            case is_empty(State) of
+                true -> mark_indexes_fixed_on_start(State);
```
I get it, a db with no indexes has no broken indexes, sometimes I'm pretty dumb, took me a few minutes to grok this one.

russelldb commented on src/riak_kv_eleveldb_backend.erl:
```diff
@@ -58,7 +62,9 @@
-                fold_opts = [{fill_cache, false}]
+                fold_opts = [{fill_cache, false}],
+                fixed_indexes = false, %% true if legacy indices have be rewritten
+                legacy_indices = false %% true if new writes use legacy indices (downgrade)
```
why is this `indices` and the previous `indexes`?

jrwest replied:
bad judgement on my part. i noticed the same thing with `mark_indexes_fixed` and `set_legacy_indices`. will change (will hold off until all review comments tho).

russelldb commented on src/riak_kv_eleveldb_backend.erl:
```diff
@@ -159,13 +191,76 @@ put(Bucket, PrimaryKey, IndexSpecs, Val, #state{ref=Ref,
+            {error, Reason, State}
+    end.
+
+set_legacy_indices(State, WriteLegacy) ->
```
yeah, what you said about consistent naming

jrwest committed:
indices -> indexes

russelldb commented:
Tests pass. Poking, noodling and smoke testing on my local cluster all works fine. You've addressed the naming comments so +1 from me. Nice work.

engelsanchez commented:
The results of this branch have been squashed into #516 
That PR will be merged and this will be closed and left behind as a detailed trail of what went on here.
